c10/core/TensorImpl.h

```diff
@@ -860,6 +803,20 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
     return impl;
   }
 
+  // NOTE: `shallow_copy_from()` does not copy the AutogradMeta pointer
+  // because it is unique for each Variable.
+  virtual void shallow_copy_from(c10::intrusive_ptr<TensorImpl> impl) {
+    type_id_ = impl->type_id();
+    is_variable_ = impl->is_variable();
+    set_storage(impl->storage());
+    set_sizes_and_strides(impl->sizes(), impl->strides());
+    set_storage_offset(impl->storage_offset());
+    is_wrapped_number_ = impl->is_wrapped_number_;
+    reserved_ = impl->reserved_;
+    refresh_numel();
+    refresh_contiguous();
+  }
```

Comment:
This method (and the one in SparseTensorImpl.h) is necessary for `variable.set_data(data_tensor)` to work, because the new implementation of `variable.set_data(data_tensor)` is to shallow-copy all TensorImpl fields (except AutogradMeta pointer) from `data_tensor` to `variable`, so that the semantics of `set_data()` can be preserved (i.e. setting the data of a Variable without changing its autograd fields).

Reply:
Why is this method virtual?

Reply:
Update: Added `NOTE [ TensorImpl Shallow-Copying ]` to TensorImpl.h to explain why we need `shallow_copy_from()`